jxl/src/frame.rs

```diff
@@ -207,28 +294,34 @@ mod test {
         util::test::assert_almost_eq,
     };
 
-    use super::{Frame, Section};
+    use super::{DecoderState, Frame, Section};
 
-    fn read_frames(image: &[u8]) -> Result<Vec<Frame>, Error> {
+    fn read_frames(
+        image: &[u8],
+        mut callback: impl FnMut(Frame) -> Result<Option<DecoderState>, Error>,
+    ) -> Result<(), Error> {
         let codestream = ContainerParser::collect_codestream(image).unwrap();
         let mut br = BitReader::new(&codestream);
         let file_header = FileHeader::read(&mut br).unwrap();
-        let mut frames = vec![];
+        let mut decoder_state = DecoderState::new(file_header);
         loop {
-            let mut frame = Frame::new(&mut br, &file_header)?;
-            let is_last = frame.is_last();
+            let mut frame = Frame::new(&mut br, decoder_state)?;
             let mut sections = frame.sections(&mut br)?;
             frame.decode_lf_global(&mut sections[frame.get_section_idx(Section::LfGlobal)])?;
-            frames.push(frame);
-            if is_last {
+
+            // Call the callback with the frame
+            if let Some(state) = callback(frame)? {
```

Comment:
I'm not sure I like the interface her - I think `callback(frame_index, &frame)` would likely work better (and should still let us do the same checks), but no strong opinion for now

Reply:
ok, then let's leave it for now...